app/services/user-state/index.ts

```diff
@@ -0,0 +1,213 @@
+
+      this.state.stateFlat = { ...this.state.stateFlat, ...changes };
+    } catch (e: unknown) {
+      this.log('error updating state:', e);
+      throw e;
+    }
+  }
+
+  async getUserState(keys: string[]) {
+    this.log('getUserState()', { keys });
+
+    if (!this.state.stateFlat) {
+      return undefined;
+    }
+
+    // filter stateFlat to only include keys in "keys"
+    return Object.fromEntries(
+      Object.entries(this.state.stateFlat).filter(([key]) => keys.includes(key)),
+    );
+  }
+
+  private onSourceMessageCallback = async (evt: { sourceName: string; message: any }[]) => {
+    for (const { sourceName, message } of evt) {
+      const sourceView = this.sourcesService.views.getSource(sourceName);
+      const source = sourceView?.getObsInput();
+
+      if (!source || !sourceView) {
+        this.log(`Source ${sourceName} not found`);
+        continue;
+      }
+
+      const parsed = JSON.parse(message);
+
+      if ('type' in parsed) {
+        switch (parsed.type) {
+          case 'getState': {
+            const key = sourceView.sourceId;
+
+            this.log({
+              type: parsed.type,
+              keys: parsed.keys,
+              sourceName,
+              sourceId: key,
+            });
+
+            this.sourceStateKeyInterest.set(
+              key,
+              new Set([...(this.sourceStateKeyInterest.get(key) ?? []), ...parsed.keys]),
+            );
+
+            this.log(key, 'using state keys: ', this.sourceStateKeyInterest.get(key));
+
+            const keys = parsed.keys;
+
+            const s = await this.getUserState(keys);
+
+            if (!s) {
+              this.log('No state available yet, skipping response to source');
+              return;
+            }
+
+            const message = toDotNotation(s);
+
+            const payload = JSON.stringify({
+              type: 'state.update',
+              message,
+              key: keys?.join(','),
+              event_id: uuid(),
+            });
+
+            source.sendMessage({ message: payload });
+          }
+        }
+      }
+
+      this.log(`unhandled source message from ${sourceName}:`, parsed);
+    }
+  };
+
+  private async fetchFullState(): Promise<StateTree> {
+    return (await this.authedRequest(
+      `https://${this.hostsService.streamlabs}/api/v5/user-state/desktop/query`,
+      { body: JSON.stringify({ query: {} }) },
+    )) as StateTree;
+  }
+
+  private async authedRequest(
+    url: string,
+    {
+      body,
+      timeoutMs = 8_000,
+      method = 'POST',
+    }: {
+      body: string;
+      timeoutMs?: number;
+      method?: 'POST' | 'PUT';
+    },
+  ) {
+    const headers = authorizedHeaders(
+      this.userService.apiToken,
+      new Headers({ 'Content-Type': 'application/json' }),
+    );
+
+    const controller = new AbortController();
+    const { signal } = controller;
+    const timeout = setTimeout(() => controller.abort(), timeoutMs);
```

Comment:
not sure why we have a timeout on our requests here? is this service expected to timeout often?